Source/WebCore/xml/XMLHttpRequest.cpp

```diff
@@ -930,6 +930,13 @@ void XMLHttpRequest::didFinishLoading(ResourceLoaderIdentifier, const NetworkLoa
     if (m_error)
         return;
 
+    // Make sure that didSendData() was called at least one before marking the load as complete
+    // so that a progress events get fired on m_upload.
+    if (m_uploadListenerFlag && m_requestEntityBody && !m_wasDidSendDataCalled) {
```

Comment:
This logic ensures we'll always send at least one progress event. But it does not ensure, after we've sent one, that we'll always send a progress event that covers the last chunk of bytes. I wonder what the spec actually requires here. Maybe we do want to always cover the last chunk of bytes.

Reply:
Seems worth checking the spec and/or other browser about this.